### PR TITLE
Resolving NPE and adding a unit test to confirm expected behavior

### LINKED ISSUE #27249: PasswordUtil throws NullPointerException on certain input
**Describe the bug**  

PasswordUtil, called from any webApplication using passwordUtil-1.0 feature, throws a `NullPointerException` on certain input.

Example input: `{redacted}mystring{/redacted}`

**Steps to Reproduce**

* Use the feature passwordUtil-1.0
* Call `PasswordUtil.passwordEncode(password, "aes")` where  `password` variable is set to something like `{redacted}mystring{/redacted}`.

**Expected behavior**  

The String `{redacted}mystring{/redacted}` will be aes-encoded with the default key.

Hard to say without a javadoc or source in maven central... is there a method contract?
According to these lines, this behaviour is unexpected:

https://github.com/OpenLiberty/open-liberty/blob/4c5ff746c397155eccb87365b7cacec134966e48/dev/com.ibm.ws.crypto.passwordutil/src/com/ibm/websphere/crypto/PasswordUtil.java#L440-L449

I think the bug is among these lines:

https://github.com/OpenLiberty/open-liberty/blob/4c5ff746c397155eccb87365b7cacec134966e48/dev/com.ibm.ws.crypto.passwordutil/src/com/ibm/websphere/crypto/PasswordUtil.java#L475-L484

In the else branch, there is a missing call to `isValidCryptoAlgorithm()`:

```
        } else if (current_crypto_algorithm != null) {
            decoded_string = passwordDecode(decoded_string);
        }
```

**Diagnostic information:**  
 - OpenLiberty Version: any from 22.0.0.6 onwards
 - Affected feature(s): passwordUtil-1.0 (possibly 1.1, did not try)
 - Java Version: Tested on Java 21 m0 and m1, as well as 17 (all Semeru), but it is a coding issue in the passwordUtil implementation.
 - server.xml configuration (WITHOUT sensitive information like passwords): n/a, just the feature
 - If it would be useful, upload the messages.log file found in `$WLP_OUTPUT_DIR/messages.log`

```
[2024-01-02T16:52:34.827+0100] 00000053 SystemOut                                                    O 2024-01-02T16:52:34,825 [Default Executor-thread-20] ERROR d.f.d.c.o.w.WlpDataTreePasswordEncrypter - Unable to encrypt password for property Type [datasource].
java.lang.NullPointerException: Cannot invoke "java.lang.String.trim()" because "decoded_string" is null
	at com.ibm.websphere.crypto.PasswordUtil.passwordEncode(PasswordUtil.java:487) ~[com.ibm.ws.crypto.passwordutil_1.0.78.jar:?]
	[internal classes]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:408) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161) ~[io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364) [io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164) [io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247) [io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:255) [io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:60) [io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:55) [io.openliberty.org.jboss.resteasy.common.jakarta_1.0.78.jar:?]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587) [io.openliberty.jakarta.servlet.5.0_1.0.78.jar:?]
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1260) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:748) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:445) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer.filter.WebAppFilterChain.invokeTarget(WebAppFilterChain.java:197) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer.filter.WebAppFilterChain.doFilter(WebAppFilterChain.java:100) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.security.jaspi.JaspiServletFilter.doFilter(JaspiServletFilter.java:58) [io.openliberty.security.jaspic.2.0.internal_1.0.78.jar:?]
	at com.ibm.ws.webcontainer.filter.FilterInstanceWrapper.doFilter(FilterInstanceWrapper.java:203) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer.filter.WebAppFilterChain.doFilter(WebAppFilterChain.java:93) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.doFilter(WebAppFilterManager.java:1068) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1263) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1077) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:77) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer40.servlet.CacheServletWrapper40.handleRequest(CacheServletWrapper40.java:87) [com.ibm.ws.webcontainer.servlet.4.0.jakarta_1.0.78.jar:?]
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:969) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:283) [com.ibm.ws.webcontainer.jakarta_1.1.78.jar:?]
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1248) [com.ibm.ws.transport.http_1.0.78.jar:?]
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:470) [com.ibm.ws.transport.http_1.0.78.jar:?]
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:429) [com.ibm.ws.transport.http_1.0.78.jar:?]
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:569) [com.ibm.ws.transport.http_1.0.78.jar:?]
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:503) [com.ibm.ws.transport.http_1.0.78.jar:?]
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:363) [com.ibm.ws.transport.http_1.0.78.jar:?]
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:330) [com.ibm.ws.transport.http_1.0.78.jar:?]
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallb
```

**Additional context**  

IBM case TS015097660

### PR DESCRIPTION
**Pull request by mshah0722**
Resolving the Null Pointer Exception brought up in #27249

### CODE REVIEW

File: dev/com.ibm.ws.crypto.passwordutil/src/com/ibm/websphere/crypto/PasswordUtil.java [hunk 1]
```diff
@@ -481,10 +481,15 @@ public static String passwordEncode(String decoded_string, String crypto_algorit
             return null;
         } else if (current_crypto_algorithm != null) {
             decoded_string = passwordDecode(decoded_string);
```
Comment (Zech-Hein) [inline dev/com.ibm.ws.crypto.passwordutil/src/com/ibm/websphere/crypto/PasswordUtil.java]: It is better performance to put the check from line 486 before the passwordDecode on line 483

**Comment by mshah0722** [comment]
#libby
#build

**Review by utle — COMMENTED** [review]
You need to update the copyright to 2024.

File: dev/com.ibm.ws.crypto.passwordutil/src/com/ibm/websphere/crypto/PasswordUtil.java [hunk 2]
```diff
@@ -480,11 +480,20 @@ public static String passwordEncode(String decoded_string, String crypto_algorit
                 return decoded_string.trim();
             return null;
         } else if (current_crypto_algorithm != null) {
+            // Return Null if the decoded_string is not tagged with a valid crypto algorithm to avoid NPE.
+            if (!isValidCryptoAlgorithm(current_crypto_algorithm)) {
+                // If it contains an invalid "{}", return a debug message.
+                if (current_crypto_algorithm.contains("{") && current_crypto_algorithm.contains("}")){
```
Comment (Zech-Hein) [inline dev/com.ibm.ws.crypto.passwordutil/src/com/ibm/websphere/crypto/PasswordUtil.java]:
```suggestion
                // The {<crypto algorithm>} tag must start at the beginning of the password(decoded_string) string
                if (current_crypto_algorithm.indexOf('{') == 0 && current_crypto_algorithm.indexOf('}') > 0){
```

**Review by Zech-Hein — CHANGES_REQUESTED** [review]

**Comment by mshah0722** [comment]
#libby
#build

**Review by bmarwell — APPROVED** [review]
LGTM, exactly the change I had in mind! Thanks for bringing this up!

**Comment by mshah0722** [comment]
#libby
#build

**Review by Zech-Hein — APPROVED** [review]

**Comment by Zech-Hein** [comment]
Fixes #PH59416

**Pull request merged**